include/gsl/gsl

```diff
@@ -74,69 +74,65 @@ class not_null
     static_assert(std::is_assignable<T&, std::nullptr_t>::value, "T cannot be assigned nullptr.");
 
 public:
-    not_null(T t) : ptr_(t) { ensure_invariant(); }
-    not_null& operator=(const T& t)
-    {
-        ptr_ = t;
-        ensure_invariant();
-        return *this;
-    }
+    constexpr not_null(T t) : ptr_(t) { ensure_invariant(); }
 
     not_null(const not_null& other) = default;
     not_null& operator=(const not_null& other) = default;
 
     template <typename U, typename Dummy = std::enable_if_t<std::is_convertible<U, T>::value>>
-    not_null(const not_null<U>& other)
+    constexpr not_null(const not_null<U>& other) : ptr_(other.get())
     {
-        *this = other;
+        ensure_invariant(); // Just in case a non-null U converts to null T.
     }
 
-    template <typename U, typename Dummy = std::enable_if_t<std::is_convertible<U, T>::value>>
-    not_null& operator=(const not_null<U>& other)
+    constexpr T get() const
     {
-        ptr_ = other.get();
-        return *this;
+        // help the optimizer
+        GSL_ASSUME(ptr_ != nullptr);
```

Comment:
Aside: This is pre-existing (See line 110 red below). I didn't add the assume, I simply made it work on GCC and Clang as well as VC++.

It would be pointless to incorporate an assumption into `Ensures`: `Ensures` checks the condition, and optimizers already recognize that `if (!foo) { throw XXX; /* or */ call_no_return_function(); }` ensures that `foo` holds in the following.

I agree with the original code's choice to use assumption rather than verification here. In my experience, compilers are fairly stupid about recognizing that invariants established by all the constructors and modifying member functions must hold in the non-modifying members. A library component like this is only useful when the abstraction cost is extremely low, and I think it will fail to meet that bar if we force the compiler to revalidate the class invariant on every access for which it fails to prove that invariant holds.

I'm not a big fan of postcondition validation in general: it's great when you're developing and validating a new component, but destroys zero-cost abstractions when applied to library components. Users don't want to pay debugging costs for our libraries, they want our libraries to work and get out of the way so they can debug their own code.